pkg/admissioncontroller/webhooks/auth/seed/graph/debug.go

```diff
@@ -0,0 +1,186 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	gonumgraph "gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/simple"
+)
+
+// DebugHandlerPath is the HTTP handler path for this debug handler.
+const DebugHandlerPath = "/debug/resource-dependency-graph"
+
+type handler struct {
+	graph *simple.DirectedGraph
+}
+
+// NewDebugHandler creates a new HTTP handler for debugging the resource dependency graph.
+func NewDebugHandler(graph *graph) http.HandlerFunc {
+	return (&handler{graph.graph}).Handle
+}
+
+func (h *handler) Handle(w http.ResponseWriter, r *http.Request) {
+	var (
+		out string
+
+		kind      = getQueryParameter(r.URL.Query(), "kind")
+		namespace = getQueryParameter(r.URL.Query(), "namespace")
+		name      = getQueryParameter(r.URL.Query(), "name")
+
+		nodesIterator = h.graph.Nodes()
+		nodes         []*vertex
+	)
+
+	// On large landscapes, if there are many vertices then the entire graph cannot be rendered fast enough. Hence, we
+	// apply some default filtering for the 'seed' kind.
+	if kind == "" && namespace == "" && name == "" && nodesIterator.Len() > 2000 {
+		kind = "Seed"
+	}
+
+	// Filter for all relevant nodes and sort them.
+	for nodesIterator.Next() {
+		v := nodesIterator.Node().(*vertex)
+
+		if (kind != "" && vertexTypes[v.vertexType] != kind) ||
+			(namespace != "" && v.namespace != namespace) ||
+			(name != "" && v.name != name) {
+			continue
+		}
+
+		nodes = append(nodes, v)
+	}
+	sort.Sort(vertexSorter(nodes))
+
+	// Render filtering form.
+	out += `
+<form action="` + DebugHandlerPath + `" method="GET">
+  <select name="kind">`
+	out += fmt.Sprintf(`<option value=""%s>&lt;all&gt;</option>`, selected("", kind))
+	for _, vt := range vertexTypes {
+		out += fmt.Sprintf(`<option value="%s"%s>%s</option>`, vt, selected(vt, kind), vt)
+	}
+	out += `
+  </select>
+  <input type="test" name="namespace" value="` + namespace + `" />
+  <input type="test" name="name" value="` + name + `" />
+  <input type="submit" value="go" />
+</form>` + separate(true)
+
+	// Iterate over each nodes, determine and sort their incoming and outgoing neighbors, and render the output.
+	for i, v := range nodes {
+		out += indent(0, "# %s", link(v))
+
+		for _, n := range []struct {
+			prefix   string
+			iterator gonumgraph.Nodes
+		}{
+			{"<-", h.graph.To(v.ID())},
+			{"->", h.graph.From(v.ID())},
```

Comment:
```suggestion
			{"<- (outgoing)", h.graph.To(v.ID())},
			{"-> (incoming)", h.graph.From(v.ID())},
```
To clarify the meaning of the arrows.
```suggestion
			{"<-", h.graph.To(v.ID())},
			{"->", h.graph.From(v.ID())},
```

Reply:
OK, will add it, but it's vice-versa: `<-` is `incoming` and `->` is `outgoing`. 😀